### DISCUSSION TITLE
help wanted

### DISCUSSION THREAD

**Discussion by pljakobs** [SmingHub/Sming#2652] [General]
dear sming community,

I have inherited a firmware project built on a rather old version of sming and am struggeling porting it forward to the current sming version. 
On the way there, I have written a very rudimentary hardware pwm support fort for sming, but going through the process of understanding the differences throughout the versions is really tough as I have no previous experience with sming. 
If anyone has the time and interest to help out, I'd be greatly appreciative.

here's the project: https://github.com/verybadsoldier/esp_rgbww_firmware

thanks,

pj

**Comment by mikee47** [comment]
Not complete but see https://github.com/mikee47/esp_rgbww_firmware/tree/feature/sming5

**Comment by pljakobs** [comment]
ok, wow, that's clearly not what I have expected! Thank you! I really don't know what to say.

Am I right in thinking that the OTA update will have to be rewritten for the esp32(c3) SOCs? 
Other than that, the only thing I can see is still a signed/unsigned issue on the c3. I'll open a pr for that. (this should be this, I think? https://github.com/espressif/esp-idf/issues/9511 - for now, I just do a static cast for risc-v 32 bit architecture)

pj

**Comment by mikee47** [comment]
The signed/unsigned can be converted to a warning  by building with `make STRICT=1`

**Comment by pljakobs** [comment]
so, I did fiddle some more with the partition layout (the original fw used two spiffs partitions, I don't think that's necesary, but I'll keep it for now).
Now, I get the system to compile all right for the ESP8266, but fail to bring up the web ui.
I assume that it's a memory issue, as I can serve smaller pages up to 2.5kB but the 6.7kB init file won't load, the client sends a get, the server prints the debug that it's about to send that file and that's it. The ESP doesn't crash, though, it will still happily serve other, smaller pages after that, just the large ones won't. 

This is the debug output:

```
[Components/Network/src/Network/Http/HttpConnection.cpp:127] The headers are complete
[/home/pjakobs/devel/esp_rgbww_firmware/app/webserver.cpp:259] entered webserver request
[/home/pjakobs/devel/esp_rgbww_firmware/app/webserver.cpp:286] not connected to ap, serving init.html
[Core/Data/Stream/IFS/FileStream.cpp:31] attached file: 'init.html.gz' (2693 bytes) #0x3FFF3E20
[Components/Network/src/Network/Http/HttpResponse.cpp:72] found init.html.gz
[Core/Data/Stream/MemoryDataStream.cpp:34] MemoryDataStream::realloc 0 -> 17
[Components/Network/src/Network/TcpClient.cpp:112] Sending stream. Bytes to send: 17
[Core/Data/Stream/MemoryDataStream.cpp:34] MemoryDataStream::realloc 17 -> 169
[Components/Network/src/Network/TcpClient.cpp:112] Sending stream. Bytes to send: 41
[Components/Network/src/Network/TcpClient.cpp:112] Sending stream. Bytes to send: 66
[Components/Network/src/Network/TcpClient.cpp:112] Sending stream. Bytes to send: 88
[Components/Network/src/Network/TcpClient.cpp:112] Sending stream. Bytes to send: 110
[Components/Network/src/Network/TcpClient.cpp:112] Sending stream. Bytes to send: 134
[Components/Network/src/Network/TcpClient.cpp:112] Sending stream. Bytes to send: 160
[Components/Network/src/Network/TcpClient.cpp:112] Sending stream. Bytes to send: 162
[Components/Network/src/Network/TcpConnection.cpp:198] TCP 3fff6468 onReadyToSendData: 1
[Components/Network/src/Network/TcpConnection.cpp:256] TCP 3fff6468 Written: 162, Available: 2780, isFinished: 0, PushCount: 1
[Components/Network/src/Network/TcpClient.cpp:198] TcpClient stream finished
[Components/Network/src/Network/TcpConnection.cpp:256] TCP 3fff6468 Written: 1024, Available: 2618, isFinished: 0, PushCount: 1
[Components/Network/src/Network/TcpConnection.cpp:256] TCP 3fff6468 Written: 1024, Available: 1594, isFinished: 0, PushCount: 2
[Components/Network/src/Network/TcpConnection.cpp:256] TCP 3fff6468 Written: 570, Available: 570, isFinished: 0, PushCount: 3
[Components/Network/src/Network/TcpConnection.cpp:136] TCP 3fff6468 received: 381 bytes
```
many seconds later, this is the next Network message:
```
[Components/Network/src/Network/TcpConnection.cpp:158] TCP 3fff6468 connection closed by timeout: 5 (from 5)
[Components/Network/src/Network/TcpConnection.cpp:289] TCP 3fff6468 connection closing
[Components/Network/src/Network/TcpConnection.cpp:36] TCP 3fff6468 ~connection
```
Looking at Wireshark, the esp ACKs the get request, reporting a window of 5179 - which seems reasonable and fits the fact that small files are being served, large ones fail. After that, I get three keepalive requests from the browser that are ACK'd. After that, the server resets the connection.

No data is ever received by the client. 

Any idea would be helpful. 

btw: the actual tcp code uses "debug_tcp_d" rather than "debug_d" - is there a build switch to get that activated?

**Comment by pljakobs** [comment]
ventured a bit into gdb. Not sure how valid this information is, due to gdb stub ram useage as well as the connection having long timed out as I'm stepping through this, but:

```
253                     ++pushCount;
(gdb) n
255                     int bytesWritten = write(buffer, bytesRead, TCP_WRITE_FLAG_COPY | TCP_WRITE_FLAG_MORE);
(gdb) n
259                     if(bytesWritten < 0) {
(gdb) print bytesWritten
$11 = 1024
(gdb) n
263                     if(bytesWritten == 0) {
(gdb) n
267                     total += size_t(bytesWritten);
(gdb) n
268                     stream->seek(bytesWritten);
(gdb) n

Breakpoint 2, TcpConnection::write (this=0x3fff6850, this@entry=0x3fff4228, stream=0x3fff4228) at /home/pjakobs/devel/esp_rgbww_firmware/Sming/Sming/Components/Netwo
   rk/src/Network/TcpConnection.cpp:241
241             while((tcp_sndqueuelen(tcp) < TCP_SND_QUEUELEN) && !stream->isFinished() && (pushCount < 25)) {
(gdb) n
242                     size_t available = getAvailableWriteSize();
(gdb) n
243                     if(available == 0) {
(gdb) print available
$12 = 570
(gdb) n
248                     auto bytesRead = stream->readMemoryBlock(buffer, std::min(sizeof(buffer), available));
(gdb) n
249                     if(bytesRead == 0) {
(gdb) n
253                     ++pushCount;
(gdb) n
255                     int bytesWritten = write(buffer, bytesRead, TCP_WRITE_FLAG_COPY | TCP_WRITE_FLAG_MORE);
(gdb) print byteswritten
No symbol "byteswritten" in current context.
(gdb) print bytesWritten
$13 = <optimized out>
(gdb) n
259                     if(bytesWritten < 0) {
(gdb) print bytesWritten
$14 = 570
(gdb) n
263                     if(bytesWritten == 0) {
(gdb) n
267                     total += size_t(bytesWritten);
(gdb) n
268                     stream->seek(bytesWritten);
(gdb) n

Breakpoint 2, TcpConnection::write (this=0x3fff6850, this@entry=0x3fff4228, stream=0x3fff4228) at /home/pjakobs/devel/esp_rgbww_firmware/Sming/Sming/Components/Netwo
   rk/src/Network/TcpConnection.cpp:241
241             while((tcp_sndqueuelen(tcp) < TCP_SND_QUEUELEN) && !stream->isFinished() && (pushCount < 25)) {
(gdb) n
242                     size_t available = getAvailableWriteSize();
(gdb) n
243                     if(available == 0) {
(gdb) print available
$15 = 0
(gdb)
```

Line TcpConnection.cpp calls getAvialableWriteSize() and that seems to go from 2670 to 0 over four calls. 
Granted, with the connection lost, no data can be sent, so it would queue up. 

I have added a debug_i to look at "available" and built without GDB and got this: 
```
[Components/Network/src/Network/TcpConnection.cpp:243] availableWriteSize: 2780
[Components/Network/src/Network/TcpConnection.cpp:243] availableWriteSize: 2618
[Components/Network/src/Network/TcpConnection.cpp:243] availableWriteSize: 1594
[Components/Network/src/Network/TcpConnection.cpp:243] availableWriteSize: 570
[Components/Network/src/Network/TcpConnection.cpp:243] availableWriteSize: 0

```
So it looks like filling up the write buffer and not freeing it (either due to a missing free or due to not sending bytes, which seems more in line with the fact that I'm not receiving bytes) is what is happening.

[edit] checking tcp_write return always comes back "0 ERR_OK" but once the availableWriteSize is down to 0, obviously, tcp_write is not called anymore as there is no available buffer.

[edit 2] I've tried to enable TCP_DEBUG und MEM_DEBUG in lwip, but building resets both lwipopts.h and user_config.h in /home/pjakobs/devel/esp_rgbww_firmware/Sming/Sming/Arch/Esp8266/Components/esp-open-lwip/esp-open-lwip/include/ - not sure why that would be the case.

**Reply by mikee47** [reply]
> [edit] checking tcp_write return always comes back "0 ERR_OK" but once the availableWriteSize is down to 0, obviously, tcp_write is not called anymore as there is no available buffer.

I guess lwip is broken somewhere...

> [edit 2] I've tried to enable TCP_DEBUG und MEM_DEBUG in lwip, but building resets both lwipopts.h and user_config.h in /home/pjakobs/devel/esp_rgbww_firmware/Sming/Sming/Arch/Esp8266/Components/esp-open-lwip/esp-open-lwip/include/ - not sure why that would be the case.

esp_open_lwip has a weird build setup... Run `make list-config | grep DEBUG` and you'll see `ENABLE_LWIPDEBUG` which does this.
